design/ratelimit-background.md

```diff
@@ -0,0 +1,219 @@
+
+
+### Kong
+
+- Supports both a local and global mode.
+- Can rate-limit by consumer, credential, ip, service, header.
+
+[Documentation](https://docs.konghq.com/hub/kong-inc/rate-limiting).
+
+### NGINX
+
+Supports annotations specifying rate limits *per client IP* for the given `Ingress`:
+- `nginx.ingress.kubernetes.io/limit-rps`
+- `nginx.ingress.kubernetes.io/limit-rpm`
+- etc.
+
+[Documentation](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#rate-limiting).
+
+### Traefik
+
+- Supports rate-limiting by source IP, header, host.
+
+[Documentation](https://doc.traefik.io/traefik/middlewares/ratelimit/).
+
+## Sketch of Possible Approaches
+
+### Local rate limiting only
+
+In this approach, Contour only exposes Envoy's local rate limiting capability, meaning each Envoy pod independently applies rate limiting to the requests it is proxying. Rate-limiting token buckets can be defined for all requests, for an entire virtual host, or for a particular route. It's not possible to define per-client rate limits.
+
+### Global rate limiting
+
+#### Simple
+
+- Support defining an RLS to use, via the `ExtensionService` API
+- Contour's API allows descriptors to be defined for virtual hosts and routes
+- Contour will provide a guide on how to deploy and configure the Lyft reference RLS implementation
+
+This approach is similar to the current approach to external auth, and to other OSS Envoy-based ingress controllers, where operation and configuration of the external service is left to the user. It provides maximum flexibility for plugging in different RLS implementations. On the down side, it means configuration needs to be applied in multiple places which is less user-friendly.
+
+#### Contour-bundled RLS
+
+- Contour's example YAML can include a manifest for the Lyft RLS, and an `ExtensionService` for it
+- Contour exposes an API for defining rate limit policies, either as new fields on `HTTPProxy`, or as an entirely new CRD (`RateLimitPolicy` or similar). Contour translates these into the Lyft RLS config file format.
+
+This is a more "batteries included" approach that simplifies the UX for folks who don't want to plug their own RLS, but comes at a cost of reduced flexibility.
+
+#### Lyft RLS operator
+
+- As an alternative to the previous option, an operator for the Lyft RLS could be written that defines a `RateLimitPolicy` CRD and translates it into the Lyft config file format. This would be independent of Contour.
+- Contour users would either use the `RateLimitPolicy` API directly to define their rate limits, or Contour could expose its own API (as new `HTTPProxy`fields or a new CRD) that would be translated into the operator's `RateLimitPolicy` API.
+
+## Questions
+
+- Are folks interested in local rate limiting, global rate limiting, or both?
```

Comment:
I don't see a good use case for local rate limiting. The goal of rate limiting is usually to protect a backend service for availability or cost reasons. Local rate limiting would require someone to distribute a global rate limit across the envoy fleet, then assume that clients are balanced evenly.

As I write this, I think that Contour potentially has enough information to do that. You'd need to build and expose the envoy fleet information internally, then dynamically adjust the rate limits.

Reply:
I agree that it's pretty limited in terms of what it can provide. BUT, it's also far easier to configure and operate. I put it out there to see if users are interested in it or not. 🤷‍♂️

Reply:
Right. One thing that this doc doesn't address is the purpose of rate limiting. What capabilities are we trying to give to operators and why? That would help to put the different options in context.